samples/matter/lock/src/credentials/credentials_data_types.h

```diff
@@ -0,0 +1,417 @@
+		size_t mSize;
+		char mValue[DOOR_LOCK_USER_NAME_BUFFER_SIZE];
+	};
+
+	Info mInfo;
+	Credentials mOccupiedCredentials;
+	Name mName;
+
+	/**
+	 * @brief Get required size for single credential entry.
+	 *
+	 * This method can be used to prepare a proper serialization buffer.
+	 *
+	 * @return size of single credential entry.
+	 */
+	constexpr static size_t RequiredBufferSize() { return sizeof(Info) + sizeof(Credentials) + sizeof(Name); }
+
+	/**
+	 * @brief fill User entry with data from EmberAfPluginDoorLockUserInfo struct.
+	 *
+	 * Filling an User entry from EmberAfPluginDoorLockUserInfo struct is possible only if
+	 * given size of EmberAfPluginDoorLockUserInfo::userName is lower than DOOR_LOCK_USER_NAME_BUFFER_SIZE
+	 * and size of EmberAfPluginDoorLockUserInfo::credentials is lower than
+	 * CONFIG_LOCK_MAX_NUM_CREDENTIALS_PER_USER.
+	 *
+	 *
+	 * @param userInfo a reference to the plugin.
+	 * @return CHIP_ERROR_INVALID_ARGUMENT if provided userInfo contains incorrect data.
+	 * @return CHIP_NO_ERROR if whole struct has been filled properly.
+	 */
+	CHIP_ERROR FillFromPlugin(EmberAfPluginDoorLockUserInfo &userInfo);
+
+	/**
+	 * @brief Convert whole object to EmberAfPluginDoorLockUserInfo structure.
+	 *
+	 * @param credentialInfo a reference to the plugin.
+	 * @return CHIP_ERROR_INTERNAL if saved user does not met requirements.
+	 * @return CHIP_NO_ERROR if conversion has been finished successfully.
+	 */
+	CHIP_ERROR ConvertToPlugin(EmberAfPluginDoorLockUserInfo &userInfo) const;
+
+	/**
+	 * @brief Serialize all fields into data buffer.
+	 *
+	 * `buffSize` must be set to at least the RequiredBufferSize size.
+	 *
+	 * @param buff buffer to store serialized data.
+	 * @param buffSize size of input buffer.
+	 * @return size_t output serialized data length, 0 value means error.
+	 */
+	size_t Serialize(void *buff, size_t buffSize);
+
+	/**
+	 * @brief Deserialize all fields from given buffer.
+	 *
+	 * `buffSize` must be set to at least the RequiredBufferSize size.
+	 *
+	 * @param buff buffer containing serialized data (by invoking serialized method).
+	 * @param buffSize size of output buffer.
+	 * @return CHIP_ERROR_BUFFER_TOO_SMALL if provided buffSize is too small.
+	 * @return CHIP_ERROR_INVALID_ARGUMENT if arguments are wrong.
+	 * @return CHIP_NO_ERROR if deserialization has been finished successfully.
+	 */
+	CHIP_ERROR Deserialize(const void *buff, size_t buffSize);
+};
+
+template <uint16_t N> struct IndexList {
+	struct __attribute__((__packed__)) List {
+		uint16_t mLength{ 0 }; /* Number of elements. */
+		uint16_t mIndexes[N] = { std::numeric_limits<uint16_t>::max() };
+	} mList;
+
+	static constexpr size_t kElementSize{ sizeof(mList.mIndexes[0]) };
+
+	/**
+	 * @brief Get required size for single IndexList entry.
+	 *
+	 * This method can be used to prepare a proper serialization buffer.
+	 *
+	 * @return size of single credential entry.
+	 */
+	constexpr static size_t RequiredBufferSize() { return sizeof(List); }
```

Comment:
```suggestion
	size_t RequiredBufferSize() { return mList.mLength * sizeof(mList.mIndexes[0]); }
```
but we can also leave the current code for the worst case.

Reply:
The thing is we use this constexpr in the application code to allocate buffers for serialization and deserialization. The only way this change would make sense would be if we allocated these buffers on the heap dynamically, I suppose. But that would require more refactoring and frankly I'm not really sure if it's better to to use stack or heap in this case.

Reply:
ok, thanks